pkg/antctl/raw/multicluster/join.go

```diff
@@ -0,0 +1,314 @@
+	ClusterID       string     `yaml:"clusterID,omitempty"`
+	Namespace       string     `yaml:"namespace,omitempty"`
+	LeaderClusterID string     `yaml:"leaderClusterID"`
+	LeaderNamespace string     `yaml:"leaderNamespace"`
+	LeaderAPIServer string     `yaml:"leaderAPIServer"`
+	TokenSecretName string     `yaml:"tokenSecretName,omitempty"`
+	TokenSecretFile string     `yaml:"tokenSecretFile,omitempty"`
+	ConfigFile      string     `yaml:"-"`
+	Secret          *v1.Secret `yaml:"-"`
+}
+
+var joinOpts *ClusterSetJoinConfig
+
+func (o *ClusterSetJoinConfig) validateAndComplete() error {
+	if o.ConfigFile != "" {
+		raw, err := ioutil.ReadFile(o.ConfigFile)
+		if err != nil {
+			return err
+		}
+		if err := yamlUnmarshall(raw, o); err != nil {
+			return err
+		}
+		if o.Kind != common.ClusterSetJoinConfigKind || o.APIVersion != common.ClusterSetJoinConfigAPIVersion {
+			return fmt.Errorf("unknown apiVersion or kind in config: %s", o.ConfigFile)
+		}
+
+		if o.TokenSecretName == "" && o.TokenSecretFile == "" {
+			o.Secret, err = unmarshallSecret(raw)
+			if err != nil {
+				return fmt.Errorf("failed to unmarshall Secret from config file: %v", err)
+			}
+		}
+	}
+
+	// The precedence order is that TokenSecretName > TokenSecretFile > JoinConfigFile.
+	if o.TokenSecretName == "" && o.TokenSecretFile != "" {
+		raw, err := ioutil.ReadFile(o.TokenSecretFile)
+		if err != nil {
+			return err
+		}
+		o.Secret, err = unmarshallSecret(raw)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshall Secret from token Secret file: %v", err)
+		}
+	}
+
+	if o.LeaderClusterID == "" {
+		return fmt.Errorf("the ClusterID of leader cluster is required")
+	}
+	if o.LeaderAPIServer == "" {
+		return fmt.Errorf("the API server of the leader cluster is required")
+	}
+	if o.TokenSecretName == "" && o.Secret == nil {
+		return fmt.Errorf("a member token Secret must be provided through the Secret name, or Secret file, or Secret manifest in the config file")
+	}
+	if o.LeaderNamespace == "" {
+		return fmt.Errorf("the leader cluster Namespace is required")
+	}
+	if o.ClusterSetID == "" {
+		return fmt.Errorf("the ClusterSet ID is required")
+	}
+	if o.ClusterID == "" {
+		return fmt.Errorf("the member ClusterID is required")
+	}
+	if o.Namespace == "" {
+		fmt.Printf("Antrea Multi-cluster Namespace is not specified. Use %s\n.", defaultMemberNamespace)
+		o.Namespace = defaultMemberNamespace
+	}
+
+	// Always set the Secret Namespace with the member cluster Multi-cluster Namespace
+	if o.Secret != nil {
+		o.Secret.Namespace = o.Namespace
+	}
+	return nil
+}
+
+func yamlUnmarshall(raw []byte, v interface{}) error {
+	d := yaml.NewDecoder(bytes.NewReader(raw))
+	d.KnownFields(false)
+	return d.Decode(v)
+}
+
+func unmarshallSecret(raw []byte) (*v1.Secret, error) {
+	decoder := yamlutil.NewYAMLOrJSONDecoder(bytes.NewReader(raw), 100)
+	secret := &v1.Secret{}
+	// We need to skip the first object in the config file which is ClusterSetJoinConfig
```

Comment:
Question - why we need to do the same for Secret file? Maybe we need to revise the comment?

Reply:
The output of Secret and join config both start with `---`. So when we unmarshall the Secret, we need to skip the first object, too. Revised the comment now.